vpr/src/route/rr_graph_storage.h

```diff
@@ -0,0 +1,578 @@
+     * - edge_switch(RRNodeId, t_edge_size)
+     *
+     * Only call these methods after partition_edges has been invoked. */
+    edge_idx_range edges(const RRNodeId& id) const {
+        return vtr::make_range(edge_idx_iterator(0), edge_idx_iterator(num_edges(id)));
+    }
+    edge_idx_range configurable_edges(const RRNodeId& id) const {
+        return vtr::make_range(edge_idx_iterator(0), edge_idx_iterator(num_edges(id) - num_non_configurable_edges(id)));
+    }
+    edge_idx_range non_configurable_edges(const RRNodeId& id) const {
+        return vtr::make_range(edge_idx_iterator(num_edges(id) - num_non_configurable_edges(id)), edge_idx_iterator(num_edges(id)));
+    }
+
+    t_edge_size num_edges(const RRNodeId& id) const {
+        return size_t(last_edge(id)) - size_t(first_edge(id));
+    }
+    t_edge_size num_configurable_edges(const RRNodeId& id) const;
+    t_edge_size num_non_configurable_edges(const RRNodeId& id) const;
+
+    // Get the first and last RREdgeId for the specified RRNodeId.
+    //
+    // The edges belonging to RRNodeId is [first_edge, last_edge), excluding
+    // last_edge.
+    //
+    // If first_edge == last_edge, then a RRNodeId has no edges.
+    RREdgeId first_edge(const RRNodeId& id) const {
+        return first_edge_[id];
+    }
+    RREdgeId last_edge(const RRNodeId& id) const {
+        return (&first_edge_[id])[1];
+    }
+
+    // Retrieve the RREdgeId for iedge'th edge in RRNodeId.
+    //
+    // This method should generally not be used, and instead first_edge and
+    // last_edge should be used.
+    RREdgeId edge_id(const RRNodeId& id, t_edge_size iedge) const {
+        RREdgeId first_edge = this->first_edge(id);
+        RREdgeId ret(size_t(first_edge) + iedge);
+        VTR_ASSERT_SAFE(ret < last_edge(id));
+        return ret;
+    }
+
+    // Get the destination node for the specified edge.
+    RRNodeId edge_sink_node(const RREdgeId& edge) const {
+        return edge_dest_node_[edge];
+    }
+
+    // Get the destination node for the iedge'th edge from specified RRNodeId.
+    //
+    // This method should generally not be used, and instead first_edge and
+    // last_edge should be used.
+    RRNodeId edge_sink_node(const RRNodeId& id, t_edge_size iedge) const {
+        return edge_sink_node(edge_id(id, iedge));
+    }
+
+    // Get the switch used for the specified edge.
+    short edge_switch(const RREdgeId& edge) const {
+        return edge_switch_[edge];
+    }
+
+    // Get the switch used for the iedge'th edge from specified RRNodeId.
+    //
+    // This method should generally not be used, and instead first_edge and
+    // last_edge should be used.
+    short edge_switch(const RRNodeId& id, t_edge_size iedge) const {
+        return edge_switch(edge_id(id, iedge));
+    }
+
+    /******************
+     * Fan-in methods *
+     ******************/
+
+    /* Init per node fan-in data.  Should only be called after all edges have
+     * been allocated */
```

Comment:
This is an expensive, O(N), operation so it should be called once you have a complete rr-graph and not called often.

Reply:
done